lib/charms/zookeeper/v0/cluster.py

```diff
@@ -0,0 +1,302 @@
+
+        return {
+            "host": host,
+            "server_string": server_string,
+            "server_id": str(server_id),
+            "unit_id": str(unit_id),
+            "unit_name": unit.name,
+            "state": state,
+        }
+
+    def _get_updated_servers(self, server_strings: Iterable[str], updated_state: str):
+        """Simple wrapper for building `updated_servers` for passing to app data updates."""
+        updated_servers = {}
+        for server in server_strings:
+            unit_id = str(int(re.findall(r"server.([1-9]+)", server)[0]) - 1)
+            updated_servers[unit_id] = updated_state
+
+        return updated_servers
+
+    def update_cluster(self) -> Dict[str, str]:
+        """Adds and removes members from the current ZK quroum.
+
+        To be ran by the Juju leader.
+
+        After grabbing all the "started" units that the leader can see in the peer relation unit data.
+        Removes members not in the quorum anymore (i.e `relation_departed`/`leader_elected` event)
+        Adds new members to the quorum (i.e `relation_joined` event).
+
+            Returns:
+                A mapping of Juju unit IDs and updated state for changed units
+                To be used in updating the app data
+                    e.g {"0": "added", "1": "removed"}
+        """
+        active_hosts = []
+        active_servers = set()
+
+        # grabs all currently 'started' units from unit data
+        # failed units will be absent
+        for unit in self.started_units:
+            active_hosts.append(self.unit_config(unit=unit)["host"])
+            active_servers.add(self.unit_config(unit=unit)["server_string"])
+
+        try:
+            zk = ZooKeeperManager(hosts=active_hosts, client_port=self.client_port)
+            zk_members = zk.server_members
+
+            # remove units first, faster due to no startup/sync delay
+            servers_to_remove = list(zk_members - active_servers)
+            zk.remove_members(members=servers_to_remove)
+
+            # sorting units to ensure units are added in id order
+            servers_to_add = sorted(active_servers - zk_members)
+            zk.add_members(members=servers_to_add)
+
+            self.status = ActiveStatus()
+
+            # extracts Juju unit ID from the changed servers
+            removed_servers = self._get_updated_servers(
+                server_strings=servers_to_remove, updated_state="removed"
+            )
+            added_servers = self._get_updated_servers(
+                server_strings=servers_to_add, updated_state="added"
+            )
+
+            return {**added_servers, **removed_servers}
+
+        # caught errors relate to a unit/zk_server not yet being ready to change
+        except (
+            MembersSyncingError,
+            MemberNotReadyError,
+            QuorumLeaderNotFoundError,
+            KazooTimeoutError,
+            UnitNotFoundError,
```

Comment:
These are wrapped+caught errors that we would want to `defer` an event for, i.e 'no troubleshooting needed, unit just needs to wait'. I wanted to avoid the lib itself deferring events, and as such needed to find a way of returning the 'please defer, something expected happened' state back to the charm, which was the intent of updating the `MaintenanceStatus` message.

Unexpected errors would raise normally and break execution on the unit, which might be desirable, not sure yet. 

Does that reasoning address your concerns here?

Reply:
Yeh - my comment was mostly concerning dev experience - like if there is an error trapped by this, it wouldn't necessarily be obvious to someone troubleshooting which call raised the exception in the first place.

Reply:
The error message is returned to the MaintenanceStatus of the unit while it defers, but I'll also log it out as `INFO`, actually I'll do that as default for all these custom 'expected' exceptions.